File: abstractions/dotnet/src/serialization/SerializationWriterProxyFactory.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;

namespace Microsoft.Kiota.Abstractions.Serialization
{
    /// <summary>
    /// Proxy factory that allows the composition of before and after callbacks on existing factories.
    /// </summary>
    public class SerializationWriterProxyFactory : ISerializationWriterFactory
    {
        /// <summary>
        /// The valid content type for the <see cref="SerializationWriterProxyFactory"/>
        /// </summary>
        public string ValidContentType { get { return _concrete.ValidContentType; } }
        private readonly ISerializationWriterFactory _concrete;
        private readonly Action<IParsable> _onBefore;
        private readonly Action<IParsable> _onAfter;
        private readonly Action<IParsable, ISerializationWriter> _onStartSerialization;
        /// <summary>
        /// Creates a new proxy factory that wraps the specified concrete factory while composing the before and after callbacks.
        /// </summary>
        /// <param name="concrete">The concrete factory to wrap.</param>
        /// <param name="onBeforeSerialization">The callback to invoke before the serialization of any model object.</param>
        /// <param name="onAfterSerialization">The callback to invoke after the serialization of any model object.</param>
        /// <param name="onStartSerialization">The callback to invoke when serialization of the entire model has started.</param>
        public SerializationWriterProxyFactory(ISerializationWriterFactory concrete,
            Action<IParsable> onBeforeSerialization,
            Action<IParsable> onAfterSerialization,
            Action<IParsable, ISerializationWriter> onStartSerialization)
        {
            _concrete = concrete ?? throw new ArgumentNullException(nameof(concrete));
            _onBefore = onBeforeSerialization;
            _onAfter = onAfterSerialization;
            _onStartSerialization = onStartSerialization;
        }
        /// <summary>
        /// Creates a new <see cref="ISerializationWriter" /> instance for the given content type.
        /// </summary>
        /// <param name="contentType">The content type for which a serialization writer should be created.</param>
        /// <returns>A new <see cref="ISerializationWriter" /> instance for the given content type.</returns>
        public ISerializationWriter GetSerializationWriter(string contentType)
        {
            var writer = _concrete.GetSerializationWriter(contentType);
            var originalBefore = writer.OnBeforeObjectSerialization;
            var originalAfter = writer.OnAfterObjectSerialization;
            var originalStart = writer.OnStartObjectSerialization;
            writer.OnBeforeObjectSerialization = (x) =>
            {
                _onBefore?.Invoke(x); // the callback set by the implementation (e.g. backing store)
                originalBefore?.Invoke(x); // some callback that might already be set on the target
            };
            writer.OnAfterObjectSerialization = (x) =>
            {
                _onAfter?.Invoke(x);
                originalAfter?.Invoke(x);
            };
            writer.OnStartObjectSerialization = (x, y) =>
            {
                _onStartSerialization?.Invoke(x, y);
                originalStart?.Invoke(x, y);
            };
            return writer;
        }
    }
}

File: abstractions/typescript/src/serialization/serializationWriterProxyFactory.ts
import { Parsable } from "./parsable";
import { SerializationWriter } from "./serializationWriter";
import { SerializationWriterFactory } from "./serializationWriterFactory";

/** Proxy factory that allows the composition of before and after callbacks on existing factories. */
export abstract class SerializationWriterProxyFactory
  implements SerializationWriterFactory {
  public getValidContentType(): string {
    return this._concrete.getValidContentType();
  }
  /**
   * Creates a new proxy factory that wraps the specified concrete factory while composing the before and after callbacks.
   * @param _concrete the concrete factory to wrap
   * @param _onBefore the callback to invoke before the serialization of any model object.
   * @param _onAfter the callback to invoke after the serialization of any model object.
   * @param _onStart the callback to invoke when the serialization of a model object starts
   */
  constructor(
    private readonly _concrete: SerializationWriterFactory,
    private readonly _onBefore?: ((value: Parsable) => void) | undefined,
    private readonly _onAfter?: ((value: Parsable) => void) | undefined,
    private readonly _onStart?:
      | ((value: Parsable, writer: SerializationWriter) => void)
      | undefined
  ) {
    if (!_concrete) {
      throw new Error("_concrete cannot be undefined");
    }
  }
  public getSerializationWriter(contentType: string): SerializationWriter {
    const writer = this._concrete.getSerializationWriter(contentType);
    const originalBefore = writer.onBeforeObjectSerialization;
    const originalAfter = writer.onAfterObjectSerialization;
    const originalStart = writer.onStartObjectSerialization;
    writer.onBeforeObjectSerialization = (value) => {
      this._onBefore && this._onBefore(value);
      originalBefore && originalBefore(value);
    };
    writer.onAfterObjectSerialization = (value) => {
      this._onAfter && this._onAfter(value);
      originalAfter && originalAfter(value);
    };
    writer.onStartObjectSerialization = (value, writer) => {
      this._onStart && this._onStart(value, writer);
      originalStart && originalStart(value, writer);
    };
    return writer;
  }
}
